keps/sig-node/4800-cpumanager-split-l3cache/README.md

```diff
@@ -0,0 +1,1170 @@
+     }
+    acc.numaOrSocketsFirst.takeFullSecondLevel()
+    if acc.isSatisfied() {
+      return acc.result, nil
+    }
+   // 2. Acquire partial uncorecache, if there are enough CPUs available to satisfy the container requirement
+   //    Acquire the full uncorecache, if available and the container requires at least all the CPUs in the uncorecache grouping
+    acc.numaOrSocketsFirst.takeThirdLevel()
+    if acc.isSatisfied() {
+        return acc.result, nil
+    }
+  ...
+  }
+  ```
+
+  - Define a new interface in the `numaOrSocketsFirstFunc` in `cpu_assignment.go` that implements `takeThirdLevel`, which calls `takeUnCoreCache`.
+
+  ```go
+  // In Split L3 Topology, we take from the sets of uncorecache as the third level
+  func (n *numaFirst) takeThirdLevel() {
+    n.acc.takeUnCoreCache()
+  }
+  // First try to take partial uncorecache (CCD), if available and the request size can fit w/in the uncorecache.
+  // Second try to take the full CCD if available and need is at least the size of the uncorecache group.
+  func (a *cpuAccumulator) takeUnCoreCache() {
+    // check if SMT ON
+    
+    for _, uncore := range a.allUnCoreCache() {
+      numCoresNeeded := a.numCPUsNeeded / a.topo.CPUsPerCore() //this is another new change
+      
+      var freeCPUsInUncorecache cpuset.CPUSet
+      //need to get needed cores in uncorecache
+      freeCoresInUncorecache := a.details.CoresNeededInUnCoreCache(numCoresNeeded, uncore)
+      klog.V(2).Info("free cores from a.details list: ", freeCoresInUncorecache)
+      for _, coreId := range freeCoresInUncorecache.List() {
+          freeCPUsInUncorecache = freeCPUsInUncorecache.Union(a.topo.CPUDetails.CPUsInCores(coreId))
+      }
+      klog.V(2).Info("freeCPUsInUncorecache  : ", freeCPUsInUncorecache)
+      if a.numCPUsNeeded == freeCPUsInUncorecache.Size() {
+          klog.V(4).InfoS("takePartialUncore: claiming cores from Uncorecache ID", uncore)
+        a.take(freeCPUsInUncorecache)
+      }
+      //take full Uncorecache if the numCPUsNeeded is greater the L3 cache size
+      a.takeFullUnCore
+      if a.isSatisfied() {
+        return
+      }
+    }
+  }
+  ```
+
+- Implement a new interface in NumaorSocket struct that checks for CPUs in a third level
+
+  ```go
+  type numaOrSocketsFirstFuncs interface {
+  takeFullFirstLevel()
+  takeFullSecondLevel()
+  takeFullThirdLevel()
+  sortAvailableL3Groups() []int
+  sortAvailableNUMANodes() []int
+  sortAvailableSockets() []int
+  sortAvailableCores() []int
+  }
+
+  // Take from the set of L3s as the third level.
+  func (n *L3GroupFirst) takeFullThirdLevel() {
+    n.acc.takeFullL3s()
+  }
+
+  ```
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
+
+HW Test Matrix
```

Comment:
I called this out in the KEP doc but it isn't clear to me what this test plan is.

I don't think you are proposing new GCP images that match this NUMA.

Reply:
We are not proposing a new GCP image to be used for the CI pipeline. The test plans listed here are unit tests for different architectures. How do you propose we could fill that in the test plan section?

Reply:
It isn't clear to me that those were unit tests.

Reply:
I agree it's not as clear, but the section covering Unit tests does not make sense to me. I added the kube coverage of what I get locally.

Reply:
This is the overall summary of how the proposed scheduling policy should behave when creating the node topology and assigning CPUs. The unit test cases will test the new functions. For e2e testing, we plan on running these pod deployments on our own baremetal instances across different architectures and confirming the CPU assignments. This feature is intended for baremetal and not cloud instances.

Reply:
Thing is, we want CI coverage in the project. We can surely start adding the e2e tests, and these tests should check the machine they are running on, and skip if the hardware is such that the feature cannot be exercised. We have already some examples in the existing tests. This is indeed suboptimal, but it's a start. The issue of actually getting cloud instances so the e2e tests can actually run is different (and AFAICT harder) and can be addressed in parallel.

Reply:
Since this is an Alpha wouldn't we just describe what want to do right? 
For e2e testing, we plan on running guaranteed pods with various CPU size requirements on our own baremetal instances across different vendor architectures and confirming the CPU assignments to L3 cache core groupings. This feature is intended for baremetal only and not cloud instances.
To cover e2e testing should cover the deployment of a pod that is following LLC alignment which can be verified in the CPU Manager State file

Reply:
We will look into how we can implement GCP instances. Added to the KEP

Reply:
> Since this is an Alpha wouldn't we just describe what want to do right?
For e2e testing, we plan on running guaranteed pods with various CPU size requirements on our own baremetal instances across different vendor architectures and confirming the CPU assignments to L3 cache core groupings. This feature is intended for baremetal only and not cloud instances.
To cover e2e testing should cover the deployment of a pod that is following LLC alignment which can be verified in the CPU Manager State file

We love that you are going to test it and its great! We are trying to catch others who may edit this work in the future and they are not aware/nor are able to test this with your system of choice.

Glad to see that we will at least have some coverage for GCP. Please feel free to ask on #sig-node or #sig-testing if you get stuck.